### PR TITLE
refactor: Rewrite exporters to use streaming API to reduce memory consumption

### PR DESCRIPTION
**Pull request by jshearer**

### CODE REVIEW

File: src/services/v1_service/service.ts [hunk 1]
```diff
@@ -74,11 +76,19 @@ export class V1Service {
...
     switch (format) {
       case "gaf":
-        const gafFile = Buffer.from(annotationsToGAF(dataset, {filters: segments.map(f=>`${f.aspect}-${f.annotationStatus}`).join(", ")}));
-        return new Return.DownloadBinaryData(gafFile,"application/csv", "gene-association.gaf");
+        const gafFileStream = annotationsToGAF(dataset, {filters: segments.map(f=>`${f.aspect}-${f.annotationStatus}`).join(", ")});
+        response.status(200);
+        response.setHeader("Content-Type", "application/csv");
+        response.setHeader("Content-disposition", "attachment;filename=gene-association.gaf");
+        gafFileStream.pipe(response);
+        return Return.NoResponse;
```
Comment (nicholastmosher) [inline src/services/v1_service/service.ts]: What does `NoResponse` mean?

File: src/utils/exporters.ts [hunk 2]
```diff
@@ -1,59 +1,95 @@
...
         header: false,
         defaultValue: "",
         excelStrings: false,
         quote: '',
         delimiter: '\t',
-    });
+    }, { objectMode: true })
+    // const serialized = serialize(data.annotations.records, );
+    //@ts-ignore
+    transformer._implicitHeader = () => {};
```
Comment (nicholastmosher) [inline src/utils/exporters.ts]: Can we get a comment for what this does?

File: src/utils/exporters.ts [hunk 3]
```diff
@@ -1,59 +1,95 @@
...
-    const header = Object.keys(additionalMetadata).length>0 ? 
-      `${data.annotations.metadata.trim()}\n${metadataSerializer(additionalMetadata).trim()}` : 
-      data.annotations.metadata
+    const header = Object.keys(additionalMetadata).length > 0 ?
+        `${data.annotations.metadata.trim()}\n${metadataSerializer(additionalMetadata).trim()}` :
+        data.annotations.metadata
 
-    return header.concat("\n",serialized);
+    const input = new Readable({ objectMode: true });
+    input._read = () => { };
```
Comment (nicholastmosher) [inline src/utils/exporters.ts]: What does this do? Can you add a comment?

File: src/utils/exporters.ts [hunk 4]
```diff
@@ -1,59 +1,95 @@
...
+
+    for (const record of data.annotations.records) {
```
Comment (nicholastmosher) [inline src/utils/exporters.ts]: Correct me if I'm wrong but this means that you're serializing _all_ of the data that's been parsed, not just a queried subset, right?

File: src/utils/exporters.ts [hunk 5]
```diff
@@ -1,59 +1,95 @@
...
+    return combine()
+        .append(header+"\n")
+        .append(transformer)
+        .append(null)
 }
 
-export function genesToCSV(data: StructuredData, additionalMetadata: {[key: string]: string} = {}){
-    const serialized = serialize(Object.values(data.genes.index).map(val=>val.gene), {
+export function genesToCSV(data: StructuredData, additionalMetadata: { [key: string]: string } = {}) {
+    const inputs = Object.values(data.genes.index).map(val => val.gene);
```
Comment (nicholastmosher) [inline src/utils/exporters.ts]: This is serializing _all_ of the genes, right?

File: src/utils/exporters.ts [hunk 6]
```diff
@@ -1,59 +1,95 @@
...
         header: true,
         defaultValue: "",
         excelStrings: false,
         quote: '',
         delimiter: '\t',
-    });
+    }, { objectMode: true });
+
+    //@ts-ignore
+    transformer._implicitHeader = () => {};
```
Comment (nicholastmosher) [inline src/utils/exporters.ts]: What is this?

File: src/utils/exporters.ts [hunk 7]
```diff
@@ -1,59 +1,95 @@
...
+
+    const header = Object.keys(additionalMetadata).length > 0 ?
+        `${data.genes.metadata.trim()}\n${metadataSerializer(additionalMetadata).trim()}` :
+        data.annotations.metadata
+
+    const input = new Readable({ objectMode: true });
+    input._read = () => { };
```
Comment (nicholastmosher) [inline src/utils/exporters.ts]: What is this?

**Comment by jshearer** [comment]
So @nicholastmosher wrt all of the weird underscore prefixed functions, they're needed for reasons unknown to me. Without them, it looks like something somewhere tries to call them and they break, but I could not figure out where or why. Streams in JS are weird...

**Pull request merged**